### PR TITLE
Allow instruction counting to use shared memory as a staging ground. (And a couple other tweaks.)

### PR DESCRIPTION
**Pull request by robieta**
This is actually something I discovered a while ago with the wall of serotonin. It was really easy for large scale runs to get bottlenecked on disk access. I have a hack in the working files of that machine to use `/dev/shm`, but I figured I should formalize and actually make a respectable utility.

I also added a param to tweak the run cadence and print when a CorePool is created; these are just to make the CI logs a bit nicer. (A printout each second on a 40 minute CI job is a bit much...)

### CODE REVIEW

File: benchmarks/instruction_counts/core/utils.py [hunk 1]
```diff
@@ -13,9 +14,8 @@
 def get_temp_dir() -> str:
     global _TEMPDIR
     if _TEMPDIR is None:
-        temp_dir = tempfile.mkdtemp()
-        atexit.register(shutil.rmtree, path=temp_dir)
-        _TEMPDIR = temp_dir
+        _TEMPDIR = _make_temp_dir(prefix="instruction_count_microbenchmarks", gc_dev_shm=True)
```
Comment (xuzhao9) [inline benchmarks/instruction_counts/core/utils.py:17]: Just curious are the temp directories cleaned up? The cleanup code seems not part of this PR?

File: benchmarks/instruction_counts/execution/runner.py [hunk 2]
```diff
@@ -43,6 +43,7 @@ def __init__(self, min_core_id: int, max_core_id: int) -> None:
         self._min_core_id: int = min_core_id
         self._max_core_id: int = max_core_id
         self._num_cores = max_core_id - min_core_id + 1
+        print(f"Core pool created: {self._num_cores} cores")
```
Comment (xuzhao9) [inline benchmarks/instruction_counts/execution/runner.py:46]: Maybe better to provide a bit more details here, I would like to know the concrete core ID list, for example:
```suggestion
        print(f"Core pool created: cores {self._min_core_id}-{self._max_core_id}")
```

File: torch/utils/benchmark/utils/common.py [hunk 3]
```diff
@@ -288,3 +293,63 @@ def set_torch_threads(n: int) -> Iterator[None]:
...
     finally:
         torch.set_num_threads(prior_num_threads)
+
+
+def _make_temp_dir(prefix: Optional[str] = None, gc_dev_shm: bool = False) -> str:
+    """Create a temporary directory. The caller is responsible for cleanup.
+
+    This function is conceptually similar to `tempfile.mkdtemp`, but with
+    the key additional feature that it will use shared memory if the
+    `BENCHMARK_USE_DEV_SHM` environment variable is set. This is an
```
Comment (xuzhao9) [inline torch/utils/benchmark/utils/common.py:303]: Why use environment variable to control the function behavior? Is it because it is hard to pass the parameters to this level?

File: torch/utils/benchmark/utils/common.py [hunk 4]
```diff
@@ -288,3 +293,63 @@ def set_torch_threads(n: int) -> Iterator[None]:
...
+        assert os.path.exists("/dev/shm"), "This system does not appear to support tmpfs (/dev/shm)."
+        os.makedirs(root, exist_ok=True)
+
+        # Because we're working in shared memory, it is more important than
+        # usual to clean up ALL intermediate files. However we don't want every
+        # worker to walk over all outstanding directories, so instead we only
+        # check when we are sure that it won't lead to contention.
+        if gc_dev_shm:
+            for i in os.listdir(root):
+                owner_file = os.path.join(root, i, "owner.pid")
```
Comment (xuzhao9) [inline torch/utils/benchmark/utils/common.py:324]: Not sure about the exact usage scenario of this file, but does it need to be protected with a file lock, if there are other processes that could access to this file?

**Review by xuzhao9 — APPROVED** [review]
Overall LGTM! Added a few questions and comments.

File: benchmarks/instruction_counts/core/utils.py (hunk 1 above)
Comment (robieta) [reply benchmarks/instruction_counts/core/utils.py:17]: It's up to the user to clean up. (Mirroring the contract of `tempfile.mkdtemp`)

File: benchmarks/instruction_counts/execution/runner.py (hunk 2 above)
Comment (robieta) [reply benchmarks/instruction_counts/execution/runner.py]: Done.

File: torch/utils/benchmark/utils/common.py (hunk 3 above)
Comment (robieta) [reply torch/utils/benchmark/utils/common.py:303]: Mainly because it's meant to be an environment level switch. (It only really matters when you're beating the hell out of a machine.)

File: torch/utils/benchmark/utils/common.py (hunk 4 above)
Comment (robieta) [reply torch/utils/benchmark/utils/common.py:324]: I considered some kind of locking scheme, but I think it's simpler to just only call it from the top level harness.

Comment (robieta) [reply torch/utils/benchmark/utils/common.py:324]: By the way, there is a check where this pass won't trigger if the PID owning a folder is still alive.
